### PR TITLE
STJ: set initial capacity of JsonObject if known (#96417)

### LINKED ISSUE #96417: [API Proposal]: Expose `JsonObject` constructor that accepts a known-length `IReadOnlyCollection` to avoid unnecessary array resize operations
### Background and motivation

[Note: this is similar to [this issue](https://github.com/dotnet/runtime/issues/91467), but that issue is for _de-_ serialization. This one is for serialization.]

**Bottom Line:** Being able to specify the capacity of `JsonPropertyDictionary` up-front would improve performance by avoiding unnecessary `Resize` calls (and resultant allocations).

**Context:** We have some high-volume serialization logic that directly encodes tuples of `(string, JsonNode)` into a `JsonNode` via the `.Add` method on the `JsonObject` type.

Because of the way that `JsonObject.IDictionary` works, its [initialization method](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.IDictionary.cs#L196) never calls the [constructor](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/JsonPropertyDictionary.cs#L29) of `JsonPropertyDictionary` that takes in a `capacity` parameter. The rub there (for our use case) is that we end up with tons of wasted `Resize` calls--because the `new List` constructor ([internal ](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/JsonPropertyDictionary.cs#L32)to `JsonPropertyDictionary`) starts as an empty array, then instantiates with a default capacity of 4, then will follow the usual double, double, double... resize pattern as-needed.  

This is wasted effort for our use case because we know up-front the number of tuples we are putting in the `JsonObject` (and it is almost always more than 4), but we have no way to "tell" the `JsonPropertyDictionary` to just start with the correct capacity to begin with. Multiply that by many thousands of API requests per minute and we quickly hit a memory allocation wall.

### API Proposal

```csharp
// in JsonObject.cs
namespace System.Text.Json.Nodes;

public sealed partial class JsonObject : JsonNode
{
    private readonly int? _initialCapacity = null;

    // update the existing public constructor that takes an `IEnumerable` to check whether it is an `ICollection`
    public JsonObject(IEnumerable<KeyValuePair<string, JsonNode?>> properties, JsonNodeOptions? options = null) : this(options)
    {
        if (properties is ICollection<KeyValuePair<string, JsonNode?>> propertiesCollection)
        {
            _initialCapacity = properties.Count;
        }

        foreach (KeyValuePair<string, JsonNode?> node in properties)
        {
            Add(node.Key, node.Value);
        }
    };
}
```

```csharp
// in JsonObject.IDictionary.cs

namespace System.Text.Json.Nodes;

public partial class JsonObject : IDictionary<string, JsonNode?>
{
    // update internals of `InitializeDictionary` to...
    private JsonPropertyDictionary<JsonNode?> InitializeDictionary()
    {
        // ... existing
        if (dictionary is null) {
            // ... existing
            dictionary = _initialCapacity.HasValue
                ? new JsonPropertyDictionary<JsonNode?>(caseInsensitive, _initialCapacity.Value) // NEW: use capacity argument if available
                : new JsonPropertyDictionary<JsonNode?>(caseInsensitive);
            // ... existing
        }
    }
}
```

### API Usage

For our specific use-case, it might look like this (in F#)

```fsharp
    /// Encode an object using a list of property names and encoded values.
    let object (values: (string * JsonNode) list) : JsonNode =
        let values = values |> List.map (fun (k, v) -> KeyValuePair(k, v))        
        JsonObject(values) // F# list implements IReadonlyCollection, so this would use the new, known-capacity constructor
```

But, frankly, the usage change _should_ be transparent to consumers. Any time a `JsonObject` is manually constructed with a known-length collection, it should know to use the known-capacity constructor.

### Alternative Designs

An alternative approach may be to alter the `JsonNodeOptions` to include some kind of object capacity value (but that feels clunkier/less transparent). 

### Risks

Unknown.

### PR DESCRIPTION
**Pull request by olo-ntaylor**
Fix #96417.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/JsonPropertyDictionary.cs
Comment (olo-ntaylor) [inline src/libraries/System.Text.Json/src/System/Text/Json/JsonPropertyDictionary.cs:19]: woops. will 🔥

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs:48]: Inlining a `throw` statement inside the method body impacts performance and is something we deliberately avoid in performance-sensitive methods. Just checking for `ICollection<T>` should suffice in my opinion, almost all collection types, including read-only collection types do implement it.

Comment (olo-ntaylor) [reply src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs:48]: Fair point about performance! I was hoping to be able to avoid the double-type-check business (and so avoid the need for that last branch). Looks like if I go back to just checking for `ICollection`, I won't need that anyway.. sorry I think maybe I misunderstood your last PR comment there.

For my own learning, would there be a way to avoid the double type checks here while still checking for both `ICollection` _and_ `IReadOnlyCollection`?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs:48]: Not really. It's for this reason that `TryGetNonEnumeratedCount` does not check for `IReadOnlyCollection` to avoid the added perf penalty.

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs:8]: I just realized that STJ can't take a dependency on System.Linq. I would recommend then to just add a simple check for `ICollection<T>` for all target frameworks.

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.IDictionary.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.IDictionary.cs:13]: It seems that this additional field could be avoided?

Comment (olo-ntaylor) [reply src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.IDictionary.cs:13]: Can you say a bit more about that? I am working under the assumption that we don't want to change the lazy-initialized `Dictionary`, which means (I _think_) that we have to be able to hold on to that known initial capacity until we actually do the initialization. Are you saying we should change the initialization logic? For instance, something like go ahead and initialize the dictionary in the constructor if we know the capacity?

Comment (olo-ntaylor) [reply src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.IDictionary.cs]: Let me know if https://github.com/dotnet/runtime/pull/96486/commits/782f43b41b7973f8e8df4d6f214aff6080de80cd is along the right direction 🙏

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.IDictionary.cs]: I think this could be simplified a bit. If you take look at what `InitializeDictionary()` does is that it checks if the `_dictionary` field has been created and otherwise populates with values from the `JsonElement` it might contain.

I think the constructor could be improved a lot if you just populated the dictionary directly, roughly:
```C#
JsonPropertyDictionary<..> dictionary = properties is ICollection<...> collection ? new(collection.Count) : new();

foreach (KeyValuePair<...> entry in properties)
{
    dictionary.Add(entry.Key, entry.Value);
}

_dictionary = dictionary; // write the populated dictionary to the field
```

Comment (olo-ntaylor) [reply src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.IDictionary.cs]: I think https://github.com/dotnet/runtime/pull/96486/commits/09b6f40ec030483fa8f60ec3af8c7f80c791a7b8 should be moving in the right direction.

This change has an (unfortunate?) side effect of having to copy the `value?.AssignParent` line from the existing `Add` method (because it depends on `Dictionary`). Let me know if you are ok with that!

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.IDictionary.cs
Comment (olo-ntaylor) [inline src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.IDictionary.cs:227]: extracted this bc it's used in the constructor now as well

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs:44]: I don't think DRY is a huge concern here, it's not a lot of code this is repeating. The benefits of not checking if `_dictionary` is populated on every `Add` operation is substantial.

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.IDictionary.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.IDictionary.cs:227]: Nit: I think this should preferably be a private static helper method that accepts a `JsonSerializerOptions?` parameter.

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs
Comment (olo-ntaylor) [reply src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs:44]: Speaking of perf, is there a performance-related reason why the internal `Dictionary` property is defined like:

```csharp
internal JsonPropertyDictionary<JsonNode?> Dictionary => _dictionary is { } dictionary ? dictionary : InitializeDictionary();
```

instead of using the null-coalescing operator?
```csharp
internal JsonPropertyDictionary<JsonNode?> Dictionary => _dictionary ?? InitializeDictionary();
```

Comment (olo-ntaylor) [reply src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs]: https://github.com/dotnet/runtime/pull/96486/commits/d9d68236cf6940805381814c6176f16a33ff660c and https://github.com/dotnet/runtime/pull/96486/commits/b01922f96cd6ed751a1b63fa08b0f9b9a4fad461

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.IDictionary.cs
Comment (olo-ntaylor) [reply src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.IDictionary.cs]: https://github.com/dotnet/runtime/pull/96486/commits/d9d68236cf6940805381814c6176f16a33ff660c

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs:41]: Can be simplified a bit, perhaps collapsing a few newlines in the expression as well
```suggestion
                    ? new(isCaseInsensitive, propertiesCollection.Count)
                    : new(isCaseInsensitive);
```

Comment (olo-ntaylor) [reply src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs:41]: Will that syntax sugar work for the net462 target? Rider is giving me a warning saying "missing type specification" (but not an error)

**Comment by olo-ntaylor** [comment]
@eiriktsarpalis thanks so much for your continued help with this PR. I am working on getting internal legal approval for the contributor license agreement.

While we're on the subject, though, I wanted to bounce another idea off you to see if it would be worth investigation: It appears that `JsonPropertyDictionary` has a [ListToDictionaryThreshold](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/JsonPropertyDictionary.cs#L16) which governs whether it uses a `List` or a `Dictionary` for the data under-the-hood. I can see that being another potentially avoidable performance hit. Namely, in this new world where we might know the capacity of the `JsonObject` up-front, and if we know it is more than that threshold (9, currently), then we could avoid creating and then immediately throwing away the internal `List` data structure and just create the `Dictionary` straight away.

Does that make sense to you?

**Comment by olo-ntaylor** [comment]
@dotnet-policy-service agree company="Olo, Inc."

**Comment by eiriktsarpalis** [comment]
> Does that make sense to you?

Possibly. You could try opening a new PR after this one is merged, although this one would probably need a bit of microbenchmarking to back its necessity.

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs:41]: It should, it's a language feature and not a framework feature.

Comment (olo-ntaylor) [reply src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs]: https://github.com/dotnet/runtime/pull/96486/commits/3499a487af30dc942d445b6048ac9d3459936ca2

**Review by eiriktsarpalis — APPROVED** [review]
This looks excellent, thank you 🎉

**Comment by olo-ntaylor** [comment]
> This looks excellent, thank you 🎉

Awesome!! Thanks so much for your help on this, @eiriktsarpalis (and @to11mtm)

**Pull request merged**
